File: app/src/main/java/com/techyourchance/multithreading/exercises/exercise9/ComputeFactorialUseCase.java
package com.techyourchance.multithreading.exercises.exercise9;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.WorkerThread;

import com.techyourchance.multithreading.common.BaseObservable;

import java.math.BigInteger;

import io.reactivex.Observable;

public class ComputeFactorialUseCase {

    public record Result(String result) { }

    private final Object LOCK = new Object();

    private int mNumberOfThreads;
    private ComputationRange[] mThreadsComputationRanges;
    private volatile BigInteger[] mThreadsComputationResults;
    private int mNumOfFinishedThreads = 0;

    private long mComputationTimeoutTime;

    private boolean mAbortComputation;

    public Observable<Result> computeFactorialAndNotify(final int argument, final int timeout) {
        return Observable.fromCallable(
                () -> {
                    initComputationParams(argument, timeout);
                    startComputation();
                    waitForThreadsResultsOrTimeoutOrAbort();
                    return processComputationResults();
                }
        );
    }

    private void initComputationParams(int factorialArgument, int timeout) {
        mNumberOfThreads = factorialArgument < 20
                ? 1 : Runtime.getRuntime().availableProcessors();

        synchronized (LOCK) {
            mNumOfFinishedThreads = 0;
            mAbortComputation = false;
        }

        mThreadsComputationResults = new BigInteger[mNumberOfThreads];

        mThreadsComputationRanges = new ComputationRange[mNumberOfThreads];

        initThreadsComputationRanges(factorialArgument);

        mComputationTimeoutTime = System.currentTimeMillis() + timeout;
    }

    private void initThreadsComputationRanges(int factorialArgument) {
        int computationRangeSize = factorialArgument / mNumberOfThreads;

        long nextComputationRangeEnd = factorialArgument;
        for (int i = mNumberOfThreads - 1; i >= 0; i--) {
            mThreadsComputationRanges[i] = new ComputationRange(
                    nextComputationRangeEnd - computationRangeSize + 1,
                    nextComputationRangeEnd
            );
            nextComputationRangeEnd = mThreadsComputationRanges[i].start - 1;
        }

        // add potentially "remaining" values to first thread's range
        mThreadsComputationRanges[0].start = 1;
    }

    @WorkerThread
    private void startComputation() {
        for (int i = 0; i < mNumberOfThreads; i++) {

            final int threadIndex = i;

            new Thread(() -> {
                long rangeStart = mThreadsComputationRanges[threadIndex].start;
                long rangeEnd = mThreadsComputationRanges[threadIndex].end;
                BigInteger product = new BigInteger("1");
                for (long num = rangeStart; num <= rangeEnd; num++) {
                    if (isTimedOut()) {
                        break;
                    }
                    product = product.multiply(new BigInteger(String.valueOf(num)));
                }
                mThreadsComputationResults[threadIndex] = product;

                synchronized (LOCK) {
                    mNumOfFinishedThreads++;
                    LOCK.notifyAll();
                }

            }).start();
        }
    }

    @WorkerThread
    private void waitForThreadsResultsOrTimeoutOrAbort() {
        synchronized (LOCK) {
            while (mNumOfFinishedThreads != mNumberOfThreads && !mAbortComputation && !isTimedOut()) {
                try {
                    LOCK.wait(getRemainingMillisToTimeout());
                } catch (InterruptedException e) {
                    return;
                }
            }
        }
    }

    @WorkerThread
    private Result processComputationResults() {
        if (mAbortComputation) {
            return new Result("Computation aborted");
        }

        BigInteger value = computeFinalResult();

        // need to check for timeout after computation of the final result
        if (isTimedOut()) {
            return new Result("Computation timed out");
        }

        return new Result(value.toString());
    }

    @WorkerThread
    private BigInteger computeFinalResult() {
        BigInteger value = new BigInteger("1");
        for (int i = 0; i < mNumberOfThreads; i++) {
            if (isTimedOut()) {
                break;
            }
            value = value.multiply(mThreadsComputationResults[i]);
        }
        return value;
    }

    private long getRemainingMillisToTimeout() {
        return mComputationTimeoutTime - System.currentTimeMillis();
    }

    private boolean isTimedOut() {
        return System.currentTimeMillis() >= mComputationTimeoutTime;
    }


    private static class ComputationRange {
        private long start;
        private final long end;

        public ComputationRange(long start, long end) {
            this.start = start;
            this.end = end;
        }
    }
}

File: app/src/main/java/com/techyourchance/multithreading/exercises/exercise10/ComputeFactorialUseCase.kt
package com.techyourchance.multithreading.exercises.exercise10

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import java.math.BigInteger

class ComputeFactorialUseCase {

    sealed class Result {
        data class Success(val factorial: BigInteger) : Result()
        data object Timeout : Result()
    }

    suspend fun computeFactorial(argument: Int, timeout: Int): Result {
        return withContext(Dispatchers.IO) {
            val numberOfThreads = if (argument < 20)
                1
            else Runtime.getRuntime().availableProcessors();
            val computationRanges = getThreadsComputationRanges(numberOfThreads, argument)

            try {
                withTimeout(timeout.toLong()) {
                    val deferredComputationResults = startComputation(computationRanges)
                    processComputationResults(deferredComputationResults.awaitAll())
                }
            } catch (e: TimeoutCancellationException) {
                Result.Timeout
            }
        }
    }

    private fun getThreadsComputationRanges(threads: Int, factorialArgument: Int): List<ComputationRange> {
        val computationRanges = mutableListOf<ComputationRange>()
        val computationRangeSize = factorialArgument / threads

        var nextComputationRangeEnd = factorialArgument.toLong()
        for (i in threads - 1 downTo 0) {
            computationRanges.add(
                ComputationRange(
                    nextComputationRangeEnd - computationRangeSize + 1,
                    nextComputationRangeEnd
                )
            )
            nextComputationRangeEnd = computationRanges.last().start - 1
        }

        // add potentially "remaining" values to first thread's range
        computationRanges[0] = ComputationRange(1, computationRanges[0].end)
        return computationRanges
    }

    private fun CoroutineScope.startComputation(computationRanges: List<ComputationRange>): List<Deferred<BigInteger>> {
        return computationRanges.map { range ->
            async(Dispatchers.IO) {
                var product = BigInteger("1")
                for (num in range.start..range.end) {
                    product = product.multiply(BigInteger(num.toString()))
                }
                product
            }
        }
    }

    private fun processComputationResults(rangeResults: List<BigInteger>): Result {
        val result = rangeResults.fold(BigInteger.valueOf(1)) { acc, x ->
            acc.multiply(x)
        }
        return Result.Success(result)
    }

    private data class ComputationRange(val start: Long, val end: Long)
}
